File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> list[str]:
        if not digits:
            return []

        phone = {
            '2': "abc", '3': "def", '4': "ghi", '5': "jkl",
            '6': "mno", '7': "pqrs", '8': "tuv", '9': "wxyz"
        }

        res = []
        path = []

        def backtrack(idx: int) -> None:
            if idx == len(digits):
                res.append(''.join(path))
                return

            for ch in phone[digits[idx]]:
                path.append(ch)
                backtrack(idx + 1)
                path.pop()  # undo choice, try next letter

        backtrack(0)
        return res

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        static const vector<string> phone = {
            "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        vector<string> res;
        string path;
        path.reserve(digits.size());

        backtrack(digits, 0, phone, path, res);
        return res;
    }

private:
    void backtrack(const string& digits, int idx, const vector<string>& phone,
                    string& path, vector<string>& res) {
        if (idx == (int)digits.size()) {
            res.push_back(path);
            return;
        }

        const string& letters = phone[digits[idx] - '0'];
        for (char ch : letters) {
            path.push_back(ch);
            backtrack(digits, idx + 1, phone, path, res);
            path.pop_back(); // undo choice, try next letter
        }
    }
};